File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken()); // 나무의 수
        int M = Integer.parseInt(st.nextToken()); // 필요한 나무 길이
        int[] trees = new int[N];
        int max = 0;
        // 주어진 나무들의 길이
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            trees[i] = Integer.parseInt(st.nextToken());
            max = Math.max(trees[i], max); // 최대 높이의 나무 확인
        }

        long low = 0, high = max;
        long result = 0;

        while (low <= high) {
            long mid = (low + high) / 2;
            long total = 0;
            for (int tree : trees) {
                if (tree > mid) total += (tree - mid);
            }
            if (total >= M) {
                result = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }

        }
        bw.write(result + " ");
        bw.flush();
        bw.close();
    }
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

input = sys.stdin.readline
N, M = map(int, input().split())
trees = list(map(int, input().split()))

# 이분 탐색을 위한 초기 범위 설정이다.
start, end = 0, max(trees)
result = 0

while start <= end:
    mid = (start + end) // 2  # 현재 절단 높이이다.
    # mid 높이에서 잘리는 나무의 총합을 계산한다.
    total = sum(tree - mid for tree in trees if tree > mid)
    
    if total >= M:  
        # 조건을 만족하므로 현재 mid 값을 저장하고,
        # 더 높은 절단 높이를 찾기 위해 start 값을 조정한다.
        result = mid
        start = mid + 1
    else:
        # 조건을 만족하지 않으므로, H를 낮추기 위해 end 값을 조정한다.
        end = mid - 1

print(result)